Source/WTF/wtf/SystemTracing.h

```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
 
 namespace WTF {
 
+#if USE(SYSPROF_CAPTURE)
+
+class SysprofAnnotator {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+
+    static void instantMark(const void*, const char *name, const char *description, ...)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, 0, getProcessName(), name, description, args);
+        va_end(args);
+    }
+
+    static void mark(const void*, Seconds timeDelta, const char *name, const char *description, ...)
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, timeDelta.microsecondsAs<int64_t>(), getProcessName(), name, description, args);
+        va_end(args);
+    }
+
+    static void beginMark(const void* pointer, const char* name, const char* description, ...)
+    {
+        std::pair<const void*, const void*> key = std::make_pair(pointer, (const void*) name);
+
+        char buffer[1024];
+        va_list args;
+        va_start(args, description);
+        vsnprintf(buffer, 1024, description, args);
+        va_end(args);
+
+        std::pair<int64_t, std::string> value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, buffer);
+
+        auto& instance = getInstance();
+
+        instance.lock.lock();
+        instance.ongoingMarks[key] = value;
+        instance.lock.unlock();
+    }
+
+    static void endMark(const void* pointer, const char *name, const char* description, ...)
+    {
+        std::pair<const void*, const void*> key = std::make_pair(pointer, (const void*) name);
+        std::optional<std::pair<int64_t, std::string>> value;
+
+        auto& instance = getInstance();
+
+        instance.lock.lock();
+        if (instance.ongoingMarks.contains(key)) {
+            value = WTFMove(instance.ongoingMarks[key]);
+            instance.ongoingMarks.erase(key);
+        }
```

Comment:
You can do the same in a more convenient way using `HashMap::take()`, which already removes the entry from the hash map and returns the value in one go:

```suggestion
        if (auto entry = instace.ongoingMarks.find(key))
            value = WTFMove(instance.ongoingMarks.take(entry));
```

Reply:
The thing here is that this is an `std::map`, not a `HashMap`. Mostly because I would have to write a hash function and I was lazy :P